File: kotlin-reference/src/main/kotlin/runtime/ParamValue.kt
package roundhouse

// The recursive params value type — the Kotlin analog of
// `runtime/crystal/param_value.cr` (`String | Hash | Array`) and
// `runtime/typescript/param_value.ts`. Ruby's untyped nested params Hash
// lowers to this closed union; `<Resource>Params.from_raw` narrows via
// `is Str` / `is Dict` at access sites.
//
// Unused by GET /articles, but included in Phase R to lock the shape the
// emitter targets for the params layer.
sealed interface ParamValue {
    @JvmInline value class Str(val value: String) : ParamValue
    @JvmInline value class Dict(val value: MutableMap<String, ParamValue>) : ParamValue
    @JvmInline value class Arr(val value: MutableList<ParamValue>) : ParamValue
}

File: swift-reference/Sources/App/runtime/ParamValue.swift
// The recursive params value type — the Swift analog of
// `runtime/crystal/param_value.cr` (`String | Hash | Array`) and
// `kotlin-reference/runtime/ParamValue.kt` (a sealed interface). Ruby's
// untyped nested params Hash lowers to this closed union; an enum with
// associated values is the idiomatic Swift spelling, and
// `<Resource>Params.from_raw` narrows via `case .str` / `case .dict`
// pattern matches at access sites.
//
// Unused by GET /articles, but included in Phase R to lock the shape the
// emitter targets for the params layer.
enum ParamValue {
    case str(String)
    case dict([String: ParamValue])
    case arr([ParamValue])
}
